src/common.cpp

```diff
@@ -1,10 +1,93 @@
 #include "common.hh"
+#include "zarr.hh"
 
 #include "platform.h"
 
 #include <cmath>
+#include <thread>
 
 namespace common = acquire::sink::zarr::common;
+common::ThreadPool::ThreadPool(size_t n_threads,
+                               std::function<void(const std::string&)> err)
+  : contexts_{ n_threads }
+{
+    if (n_threads == 0) {
+        throw std::runtime_error("Cannot create thread pool with 0 threads.");
+    }
+
+    for (auto& ctx_ : contexts_) {
+        ctx_.should_stop = false;
+        ctx_.ready = true;
+        ctx_.thread = std::thread([this, ctx = &ctx_] { thread_worker_(ctx); });
+    }
+}
+
+common::ThreadPool::~ThreadPool() noexcept
+{
+    // spin down threads
+    for (auto& ctx : contexts_) {
+        ctx.should_stop = true;
+        ctx.cv.notify_one();
+        ctx.thread.join();
+    }
+}
+
+void
+common::ThreadPool::push_to_job_queue(JobT&& job)
+{
+    std::scoped_lock lock(jobs_mutex_);
+    jobs_.push(std::move(job));
+}
+
+std::optional<common::ThreadPool::JobT>
+common::ThreadPool::pop_from_job_queue_() noexcept
+{
+    std::scoped_lock lock(jobs_mutex_);
+    if (jobs_.empty()) {
+        return std::nullopt;
+    }
+
+    auto job = jobs_.front();
+    jobs_.pop();
+    return job;
```

Comment:
Is `front()` still live after `pop()`? May need to use `swap()` to swap in a dummy value - or copy the value out before pop.

Reply:
In `thread_worker_`, I want to assume that the returned job is independent of the queue once returned from this function.

Reply:
It is still live, but not for any good reason. `jobs_.front()` returns a reference, true, but we're invoking the copy constructor here. If it will make you less uneasy, I can `move` it.